state/address.go

```diff
@@ -221,21 +221,71 @@ func (st *State) APIHostPortsForClients() ([][]network.HostPort, error) {
 
 // APIHostPortsForAgents returns the collection of API addresses that should
 // be used by agents.
+// If the controller model is CAAS type, the return will be the controller
+// k8s service addresses in cloud service.
 // If there is no management network space configured for the controller,
 // or if the space is misconfigured, the return will be the same as
 // APIHostPortsForClients.
 // Otherwise the returned addresses will correspond with the management net space.
 // If there is no document at all, we simply fall back to APIHostPortsForClients.
 func (st *State) APIHostPortsForAgents() ([][]network.HostPort, error) {
-	hp, err := st.apiHostPortsForKey(apiHostPortsForAgentsKey)
+	if hps, err := st.apiHostPortsForOperator(); err == nil {
+		// TODO(caas): add test for this once we have the replacement for Jujuconnsuite.
+		logger.Debugf("K8s controller hostports for CAAS operator %v", hps)
+		return hps, nil
+	}
+
+	hps, err := st.apiHostPortsForKey(apiHostPortsForAgentsKey)
 	if err != nil {
 		if err == mgo.ErrNotFound {
 			logger.Debugf("No document for %s; using %s", apiHostPortsForAgentsKey, apiHostPortsKey)
 			return st.APIHostPortsForClients()
 		}
 		return nil, errors.Trace(err)
 	}
-	return hp, nil
+	return hps, nil
+}
+
+func (st *State) apiHostPortsForOperator() ([][]network.HostPort, error) {
+	controllerUUID := st.ControllerUUID()
+
+	m, err := st.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	controllerSt, err := st.newStateNoWorkers(st.ControllerModelUUID())
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	defer controllerSt.Close()
+
+	controllerModel, err := controllerSt.Model()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+
+	if m.Type() != ModelTypeCAAS || controllerModel.Type() != ModelTypeCAAS {
+		// only CAAS model has operator and only k8s controller has cloud service.
+		return nil, errors.NotValidf(
+			"cloud service hostports for %q controller %q, %q model %q",
+			controllerModel.Type(), controllerUUID,
+			m.Type(), m.Name(),
+		)
+	}
+
+	controllerConfig, err := st.ControllerConfig()
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	apiPort := controllerConfig.APIPort()
+	svc, err := controllerSt.CloudService(controllerUUID)
+	if err != nil {
+		return nil, errors.Trace(err)
+	}
+	return [][]network.HostPort{
+		network.AddressesWithPort(svc.Addresses(), apiPort),
```

Comment:
We need to filter addresses to only get those which are of type CloudLocal
We discussed that we could store in []addresses Public and CloudLocal and so we need to filter. If there's only one, then we use that regardless.

Reply:
I thought it would be next PR. 
no problem, it's address related. I am adding this functionality now